File: GatoRide/backend/controllers/authController.js
import User from '../models/User.js';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import sendVerificationEmail from '../utils/mailer.js';

const signup = async (req, res) => {
  const { name, username, email, password } = req.body;

  try {
    // Check if user already exists with the same email or username
    let user = await User.findOne({ $or: [{ email }, { username }] });
    if (user) {
      return res.status(400).json({ msg: 'User with this email or username already exists' });
    }

    // Create a new user
    user = new User({ name, username, email, password });

    // Generate a verification token
    const verificationToken = jwt.sign({ email }, process.env.JWT_SECRET, { expiresIn: '1h' });
    user.verificationToken = verificationToken;

    // Save the user to the database
    await user.save();

    // Send verification email
    await sendVerificationEmail(email, verificationToken);

    res.status(201).json({ msg: 'User registered successfully. Please check your email to verify your account.' });
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server error');
  }
};

const verifyEmail = async (req, res) => {
  try {
      const { token } = req.query;

      if (!token) {
          return res.status(400).json({ msg: "Invalid or missing token" });
      }

      // Verify the token
      const decoded = jwt.verify(token, process.env.JWT_SECRET);

      // Find the user by email
      const user = await User.findOne({ email: decoded.email });

      if (!user) {
          return res.status(404).json({ msg: "User not found" });
      }

      // Mark user as verified
      user.isVerified = true;
      await user.save();

      res.status(200).json({ msg: "Email verified successfully! You can now log in." });
  } catch (error) {
      console.error(error);
      res.status(400).json({ msg: "Invalid or expired token" });
  }
};

const login = async (req, res) => {
  const { usernameOrEmail, password } = req.body;

  try {
    // Find the user by username or email
    const user = await User.findOne({
      $or: [{ username: usernameOrEmail }, { email: usernameOrEmail }],
    });

    if (!user) {
      return res.status(400).json({ msg: 'user notfound' });
    }

    // Check if the password is correct
    const salt = await bcrypt.genSalt(10);
    const input_password = await bcrypt.hash(password, salt);
    const isMatch = await bcrypt.compare(password, user.password);

    if (!isMatch) {
      console.log()
      return res.status(400).json({ msg: 'Invalid credentials' });
    }

    // Check if the user's email is verified
    if (!user.isVerified) {
      return res.status(400).json({ msg: 'Please verify your email first' });
    }

    // Generate a JWT token
    const accessToken = jwt.sign(
      { user: { id: user.id } },
      process.env.JWT_SECRET,
      { expiresIn: "15m" }
    );

    // ✅ Generate Refresh Token (expires in 7 days)
    const refreshToken = jwt.sign(
      { user: { id: user.id } },
      process.env.REFRESH_SECRET,
      { expiresIn: "7d" }
    );

    // ✅ Store Refresh Token in HTTP-Only Cookie
    res.cookie("refreshToken", refreshToken, {
      httpOnly: true,
      secure: true,
      sameSite: "Strict",
      maxAge: 7 * 24 * 60 * 60 * 1000, // 7 days
    });

    res.json({ accessToken });
  } catch (err) {
    console.error(err.message);
    res.status(500).send("Server error");
  }
};

const refreshAccessToken = async (req, res) => {
  const refreshToken = req.cookies.refreshToken;

  if (!refreshToken) {
    return res.status(401).json({ msg: "No refresh token, please login" });
  }

  try {
    const decoded = jwt.verify(refreshToken, process.env.REFRESH_SECRET);
    
    // Generate new Access Token
    const accessToken = jwt.sign(
      { user: { id: decoded.user.id } },
      process.env.JWT_SECRET,
      { expiresIn: "15m" }
    );

    res.json({ accessToken });
  } catch (err) {
    res.status(401).json({ msg: "Invalid refresh token, please login" });
  }
};

export default { signup, verifyEmail, login, refreshAccessToken };

File: GatoRide/backend/controllers/auth_controller.go
package controllers

import (
	"context"
	"encoding/json"
	"net/http"

	"github.com/go-auth/config"
	"github.com/go-auth/models"
	"github.com/go-auth/utils"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Signup with Email Verification
func Signup(w http.ResponseWriter, r *http.Request) {
	var user models.User
	err := json.NewDecoder(r.Body).Decode(&user)
	if err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	collection := config.GetCollection("users")

	// Check if email or username already exists
	var existingUser models.User
	err = collection.FindOne(context.TODO(), bson.M{"$or": []bson.M{{"email": user.Email}, {"username": user.Username}}}).Decode(&existingUser)
	if err == nil {
		http.Error(w, "Email or Username already exists", http.StatusConflict)
		return
	}

	// Hash password
	hashedPassword, _ := utils.HashPassword(user.Password)
	user.Password = hashedPassword
	user.ID = primitive.NewObjectID()
	user.IsVerified = false

	// Generate verification token
	verificationToken, _ := utils.GenerateVerificationToken(user.Email)
	user.VerificationToken = verificationToken

	// Insert user into DB
	_, err = collection.InsertOne(context.TODO(), user)
	if err != nil {
		http.Error(w, "Error creating user", http.StatusInternalServerError)
		return
	}

	// Send verification email
	err = utils.SendVerificationEmail(user.Email, verificationToken)
	if err != nil {
		http.Error(w, "Failed to send verification email", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]string{"message": "User created. Verify your email."})
}

// Verify Email
func VerifyEmail(w http.ResponseWriter, r *http.Request) {
	token := r.URL.Query().Get("token")

	email, err := utils.VerifyToken(token)
	if err != nil {
		http.Error(w, "Invalid or expired token", http.StatusBadRequest)
		return
	}

	collection := config.GetCollection("users")

	_, err = collection.UpdateOne(context.TODO(), bson.M{"email": email}, bson.M{"$set": bson.M{"is_verified": true}})
	if err != nil {
		http.Error(w, "Error verifying email", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]string{"message": "Email verified successfully."})
}

// Login User
func Login(w http.ResponseWriter, r *http.Request) {
	var creds models.User
	err := json.NewDecoder(r.Body).Decode(&creds)
	if err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	collection := config.GetCollection("users")
	var user models.User

	// Check if user exists
	err = collection.FindOne(context.TODO(), bson.M{"email": creds.Email}).Decode(&user)
	if err != nil {
		http.Error(w, "Invalid email or password", http.StatusUnauthorized)
		return
	}

	// Check password
	if !utils.CheckPasswordHash(creds.Password, user.Password) {
		http.Error(w, "Invalid email or password", http.StatusUnauthorized)
		return
	}

	// Check if user is verified
	if !user.IsVerified {
		http.Error(w, "Please verify your email before logging in.", http.StatusUnauthorized)
		return
	}

	// Generate JWT token for authentication
	token, _ := utils.GenerateJWT(user.Username)

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]string{"token": token})
}